app/vmstorage/servers/vmselect.go

```diff
@@ -249,10 +252,38 @@ func getMaxMetrics(sq *storage.SearchQuery) int {
 	maxMetrics := sq.MaxMetrics
 	maxMetricsLimit := *maxUniqueTimeseries
 	if maxMetricsLimit <= 0 {
-		maxMetricsLimit = 2e9
+		maxMetricsLimit = GetMaxMetricsLimitByResource()
 	}
 	if maxMetrics <= 0 || maxMetrics > maxMetricsLimit {
 		maxMetrics = maxMetricsLimit
 	}
 	return maxMetrics
 }
+
+var (
+	maxMetricsLimitByResource int
+	once                      sync.Once
+)
+
+// GetMaxMetricsLimitByResource returns the max metrics limit calculated by available resources.
+// The calculation is split into calculateMaxMetricsLimitByResource for unit testing.
+func GetMaxMetricsLimitByResource() int {
```

Comment:
nit: it would be better to rename it to `GetMaxUniqueTimeSeries` to avoid confusion with `search.maxMetrics` limit